### PR TITLE
[SPARK-9119] [SPARK-8359] [SQL] match Decimal.precision/scale with DecimalType

### PR DESCRIPTION
**Pull request by davies**
Let Decimal carry the correct precision and scale with DecimalType.

cc @rxin @yhuai

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala [hunk 1]
```diff
@@ -417,6 +417,10 @@ trait Row extends Serializable {
             if (!o2.isInstanceOf[Double] || ! java.lang.Double.isNaN(o2.asInstanceOf[Double])) {
               return false
             }
+          case d1: java.math.BigDecimal if o2.isInstanceOf[java.math.BigDecimal] =>
+            if (d1.compareTo(o2.asInstanceOf[java.math.BigDecimal]) != 0) {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala]: Compare BigDecimal different scale (used in tests)

**Comment by rxin** [comment]
LGTM (not super familiar with decimals though)

**Comment by davies** [comment]
Merged into master and 1.5 branch.

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala [hunk 2]
```diff
@@ -189,4 +189,17 @@ class ParquetQuerySuite extends QueryTest with ParquetTest {
...
+  test("SPARK-9119 Decimal should be correctly written into parquet") {
+    withTempPath { dir =>
+      val basePath = dir.getCanonicalPath
+      val schema = StructType(Array(StructField("name", DecimalType(10, 5), false)))
+      val rowRDD = sqlContext.sparkContext.parallelize(Array(Row(Decimal("67123.45"))))
+      val df = sqlContext.createDataFrame(rowRDD, schema)
+      df.write.parquet(basePath)
+
+      val decimal = sqlContext.read.parquet(basePath).first().getDecimal(0)
+      assert(Decimal("67123.45") === Decimal(decimal))
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala]: The decimal we read back should be `67123.45000`, right? I am wondering if we need a better `assert` at here? Actually, why it did not fail?

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala]: When we compare Decimal, the zero on the most right does not matter. If we compare them as java.math.BigDecimal, it matters.

**Pull request closed without merging**